# generator && async await

## fetch

```js
import fetch from "node-fetch";
const sampleUrl = "https://jsonplaceholder.typicode.com/users/1";
// const myFetch = (url) => fetch(url).then((user) => user.json());

// => to use appropriately
const myFetch = (url) =>
  fetch(url).then((user) => {
    console.log("111>", user);
    return user.json();
  });

myFetch(sampleUrl).then((user) => {
  console.log("111>", user);
});
```

## generator

## Promise with Generator

```js
const afterTime = (sec) => {
  if (sec < 1 || sec > 3)
    return Promise.reject(new Error("Not valid sec range!!"));
  return new Promise((resolve) => setTimeout(resolve, sec * 1000, sec));
};
```

1.Promise.then 사용

```js
afterTime(1)
  .then((res) => console.log("res>>>>>", res))
  .catch(console.error);
```

then 헬 문제가 있음<br/>

2.generator 이용

```js
function* pAfterTime(sec) {
  return yield afterTime(sec);
}
const pa = pAfterTime(1);
const n1 = pa.next();
n1.value
  .then((res) => {
    const n2 = pa.next(res);
    console.log("res>>>>>", n1.next);
  })
  .catch(console.error);
```

## async await

- Promise를 생성하고 소비하기 위한 문법적 설탕.
- 비동기 함수에서 콜백을 사용하는 대신 단순한 논리적 흐름(비동기, 직렬)을 작성.
- 즉, Promise의 then/catch/finally를 사용할 필요 없음!
- async는 Promise를 반환! await은 resolve, reject와 매핑!
- 예외(error)는 거부(reject)고, 거부(reject)는 예외(throw)다!
- 반환(성공)은 확정(resolve)이고, 완료(return)는 결과(result)이다!
